rakefiles/merge_yaml.rb

```diff
@@ -0,0 +1,28 @@
+require 'yaml'
+
+# Monkey patch the output format of the date to avoid kops failures
+#
+# YAML module dumps the date in an incorrect format, causing Kops fails
+#
+# https://github.com/ruby/ruby/blob/ruby_2_4/ext/psych/lib/psych/visitors/yaml_tree.rb#L522-L528
+#
+Psych::Visitors::YAMLTree.class_eval do
+  def format_time time
+    time.strftime("%Y-%m-%dT%H:%M:%SZ")
+  end
+end
+
+class InfraUtils
+
+  def self.merge_yaml(a,b)
+    YAML.dump(recurse_merge_hash(YAML.load(a), YAML.load(b)))
+  end
+
+  def self.recurse_merge_hash(a,b)
+    a.merge(b) do |_,x,y|
+      (x.is_a?(Hash) && y.is_a?(Hash)) ? recurse_merge_hash(x,y) : y
```

Comment:
This prioritizes right-hand items (whatever's y wins if the key is the same as an x thing). You should document that, but also, I'm curious: why not just use Hash.merge()? Functionality appears to be the same (see IRB session, below).

```irb(main):001:0> def recurse_merge_hash(a,b)
irb(main):002:1>   a.merge(b) do |_,x,y|
irb(main):003:2*     (x.is_a?(Hash) && y.is_a?(Hash)) ? recurse_merge_hash(x,y) : y
irb(main):004:2>   end
irb(main):005:1> end
=> :recurse_merge_hash
irb(main):006:0> foo = { "alpha" => 10, "bravo" => 20 }
=> {"alpha"=>10, "bravo"=>20}
irb(main):007:0> bar = { "charlie" => 30, "delta" => 40}
=> {"charlie"=>30, "delta"=>40}
irb(main):008:0> recurse_merge_hash(foo,bar)
=> {"alpha"=>10, "bravo"=>20, "charlie"=>30, "delta"=>40}
irb(main):009:0> bar = { "charlie" => 30}
=> {"charlie"=>30}
irb(main):010:0> recurse_merge_hash(foo,bar)
=> {"alpha"=>10, "bravo"=>20, "charlie"=>30}
irb(main):011:0> bar = { "charlie" => {"echo" => 50}}
=> {"charlie"=>{"echo"=>50}}
irb(main):012:0> recurse_merge_hash(foo,bar)
=> {"alpha"=>10, "bravo"=>20, "charlie"=>{"echo"=>50}}
irb(main):013:0> foo = { "alpha" => 10, "bravo" => 20, "charlie" => 50}
=> {"alpha"=>10, "bravo"=>20, "charlie"=>50}
irb(main):014:0> recurse_merge_hash(foo,bar)
=> {"alpha"=>10, "bravo"=>20, "charlie"=>{"echo"=>50}}
irb(main):015:0> recurse_merge_hash(bar, foo)
=> {"charlie"=>50, "alpha"=>10, "bravo"=>20}
irb(main):016:0> bar.merge(foo)
=> {"charlie"=>50, "alpha"=>10, "bravo"=>20}
irb(main):017:0> foo.merge(bar)
=> {"alpha"=>10, "bravo"=>20, "charlie"=>{"echo"=>50}}
irb(main):018:0> 
```

Reply:
I don't use Hash.merge() directly because if one key is matched in Y and X all the items of the destiny key will be removed, instead of only substitute the items that match. Try to use hash.merge() instead of recurse_merge_hash and you will see that test fails.

And yes, the items listed in the `custom_cluster_settings.yml.erb ` file have priority.

Reply:
OK. In that case (for both issues), would you mind adding code comments to point out what you're doing, how it's different than the normal .merge() case, and that right-hand items are expected to override left-hand items and that's acceptable for what you're doing?

That way, in three years (if it was my code, it'd be 3 months, but I'm assuming you have a better memory than me :-) ) you (or others reading your code) will be able to remember why you did this, and/or explain otherwise-unexplained behavior.

Otherwise LGTM.

Reply:
Comments added. And the test that covers this function is `test_yaml_substitution`. If the Hash.merge() function is used, that test will fail.